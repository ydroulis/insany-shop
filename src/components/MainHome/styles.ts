'use client'

import styled from 'styled-components'

export const Container = styled.main`
    padding: 0rem 10rem; 
    font-family: Inter, sans-serif
`