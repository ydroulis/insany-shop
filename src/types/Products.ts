export type Products = Product[]

export type Product = {
    id: number,
    name: string,
    description: string,
    price: number,
    image: string,
    category: string,
    stock: number,
    rating: number,
    brand?: string

}