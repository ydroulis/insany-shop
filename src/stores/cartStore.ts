import { Cart, CartProduct } from '../types/Cart'
import { createStore } from 'zustand/vanilla'

export type CartState = {
    cart: Cart
    feedback?: boolean
}

export type CartActions = {
    addProductToCart: (product: CartProduct, items?: number) => void
    removeProductFromCart: (id: number, price: number) => void
    changeItems: (id: number, items: number) => void
    showFeedback: (value: boolean) => void
}

export type CartStore = CartState & CartActions

export const defaultInitState: CartState = {
    cart: {
        content: [

        ],
        total: 0
    },
    feedback: false
}

export const createCartStore = (
    initState: CartState = defaultInitState,
) => {
    return createStore<CartStore>()((set) => ({
        ...initState,
        addProductToCart(product, items = 1) {
            set((state) => ({
                cart: {
                    content: [...state.cart.content, { ...product, items }],
                    total: state.cart.total + (product.price * items)
                }
            }))
        },
        removeProductFromCart(id, price) {
            set((state) => ({
                cart: {
                    content: state.cart.content.filter(product => product.id !== id),
                    total: state.cart.content.length === 1 ? 0 : state.cart.total - price
                }
            }))
        },
        changeItems(id, items) {
            set((state) => {
                return {
                    cart: {
                        content: state.cart.content.map(product => product.id === id ? { ...product, items } : product),
                        total: state.cart.total + (state.cart.content.find(product => product.id === id)?.price || 0) * (items - (state.cart.content.find(product => product.id === id)?.items || 0))
                    }
                }
            })
        },
        showFeedback(value) {
            set({ feedback: value });
        }
    }))
}