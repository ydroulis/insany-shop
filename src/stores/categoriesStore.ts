import { Categories } from '@/types/Categories'
import { createStore } from 'zustand/vanilla'

export type CategoriesState = {
    categories: Categories
}

export type CategoriesActions = {
    setCategories: (categories: Categories) => void
}

export type CategoriesStore = CategoriesState & CategoriesActions

export const defaultInitState: CategoriesState = {
    categories: [],
}

export const createCategoriesStore = (
    initState: CategoriesState = defaultInitState,
) => {
    return createStore<CategoriesStore>()((set) => ({
        ...initState,
        setCategories: (categories: Categories) => set({ categories }),
    }))
}