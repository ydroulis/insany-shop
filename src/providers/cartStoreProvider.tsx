'use client'

import { type ReactNode, createContext, useRef, useContext } from 'react'
import { useStore } from 'zustand'

import { type CartStore, createCartStore } from '../stores/cartStore'

export type CartStoreApi = ReturnType<typeof createCartStore>

export const CartStoreContext = createContext<CartStoreApi | undefined>(
    undefined,
)

export interface CartStoreProviderProps {
    children: ReactNode
}

export const CartStoreProvider = ({
    children,
}: CartStoreProviderProps) => {
    const storeRef = useRef<CartStoreApi | null>(null)
    if (storeRef.current === null) {
        storeRef.current = createCartStore()
    }

    return (
        <CartStoreContext.Provider value={storeRef.current}>
            {children}
        </CartStoreContext.Provider>
    )
}

export const useCartStore = <T,>(
    selector: (store: CartStore) => T,
): T => {
    const cartStoreContext = useContext(CartStoreContext)

    if (!cartStoreContext) {
        throw new Error(`useStore must be used within CartStoreProvider`)
    }

    return useStore(cartStoreContext, selector)
}
