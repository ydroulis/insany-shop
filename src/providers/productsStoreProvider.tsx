'use client'

import { type ReactNode, createContext, useRef, useContext } from 'react'
import { useStore } from 'zustand'

import { type ProductsStore, createProductsStore } from '../stores/productsStore'

export type ProductsStoreApi = ReturnType<typeof createProductsStore>

export const ProductsStoreContext = createContext<ProductsStoreApi | undefined>(
    undefined,
)

export interface ProductsStoreProviderProps {
    children: ReactNode
}

export const ProductsStoreProvider = ({
    children,
}: ProductsStoreProviderProps) => {
    const storeRef = useRef<ProductsStoreApi | null>(null)
    if (storeRef.current === null) {
        storeRef.current = createProductsStore()
    }

    return (
        <ProductsStoreContext.Provider value={storeRef.current}>
            {children}
        </ProductsStoreContext.Provider>
    )
}

export const useProductsStore = <T,>(
    selector: (store: ProductsStore) => T,
): T => {
    const productsStoreContext = useContext(ProductsStoreContext)

    if (!productsStoreContext) {
        throw new Error(`useStore must be used within ProductsStoreProvider`)
    }

    return useStore(productsStoreContext, selector)
}
