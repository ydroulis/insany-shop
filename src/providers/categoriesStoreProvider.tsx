'use client'

import { type ReactNode, createContext, useRef, useContext } from 'react'
import { useStore } from 'zustand'

import { type CategoriesStore, createCategoriesStore } from '../stores/categoriesStore'

export type CategoriesStoreApi = ReturnType<typeof createCategoriesStore>

export const CategoriesStoreContext = createContext<CategoriesStoreApi | undefined>(
    undefined,
)

export interface CategoriesStoreProviderProps {
    children: ReactNode
}

export const CategoriesStoreProvider = ({
    children,
}: CategoriesStoreProviderProps) => {
    const storeRef = useRef<CategoriesStoreApi | null>(null)
    if (storeRef.current === null) {
        storeRef.current = createCategoriesStore()
    }

    return (
        <CategoriesStoreContext.Provider value={storeRef.current}>
            {children}
        </CategoriesStoreContext.Provider>
    )
}

export const useCategoriesStore = <T,>(
    selector: (store: CategoriesStore) => T,
): T => {
    const categoriesStoreContext = useContext(CategoriesStoreContext)

    if (!categoriesStoreContext) {
        throw new Error(`useStore must be used within CategoriesStoreProvider`)
    }

    return useStore(categoriesStoreContext, selector)
}
