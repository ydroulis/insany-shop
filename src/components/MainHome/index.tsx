import React from 'react';

import * as S from './styles';
import ProductList from '../ProductList';
import { Products } from '@/types/Products';
import Filters from '../Filters';
import CategoriesList from '../CategoryList';
import { useProductsStore } from '../../providers/productsStoreProvider';
import { useCategoriesStore } from '../../providers/categoriesStoreProvider';

const MainHome: React.FC = () => {
    const { products } = useProductsStore((state => state));
    const { categories } = useCategoriesStore((state => state));

    const productList: Products = products.map(product => ({
        ...product,
        category: categories.find(category => category.id === product.category)?.name
    }))
    return (
        <S.Container
            as="main"
            role="main"
            aria-label="CatÃ¡logo de produtos"
        >
            <Filters />
            <ProductList products={productList} />
            <CategoriesList categories={categories} />
        </S.Container>
    );
}

export default MainHome;