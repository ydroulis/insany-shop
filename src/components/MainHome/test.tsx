import React, { useRef } from "react";
import { render, screen } from "../../../.jest/test-utils";
import MainHome from ".";
import { Product } from "@/types/Products";
import { CategoriesStoreContext, CategoriesStoreApi } from "../../providers/categoriesStoreProvider";
import { ProductsStoreProvider } from "../../providers/productsStoreProvider";
import { createCategoriesStore } from "../../stores/categoriesStore";

jest.mock("../Filters", () => {
    const mockFilters = () => <div data-testid="filters-mock">Filtros</div>;
    return mockFilters;
});

jest.mock("../ProductList", () => {
    const ProductList = ({ products }: { products: Product[] }) => (
        <ul data-testid="product-list-mock">
            {products.map((product) => (
                <li key={product.id}>
                    <img
                        src={product.image}
                        alt={`Imagem do produto ${product.name}`}
                    />
                    <span>{product.name}</span>
                    <span>{product.category}</span>
                    <span>R$ {product.price.toFixed(2)}</span>
                    <span>{product.stock} em estoque</span>
                    <span aria-label={`Avaliação ${product.rating}`}>⭐ {product.rating}</span>
                    <span>{product.brand}</span>
                </li>
            ))}
        </ul>
    );

    return ProductList;
});

const TestCategoriesProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
    const storeRef = useRef<CategoriesStoreApi | null>(null);

    if (!storeRef.current) {
        storeRef.current = createCategoriesStore();
        storeRef.current.setState({
            categories: [
                { id: 'livros', name: 'Livros', productCount: 1, description: '', icon: '' },
                { id: 'eletronics', name: 'Eletrônicos', productCount: 5, description: '', icon: '' },
            ],
        });
    }

    return (
        <CategoriesStoreContext.Provider value={storeRef.current}>
            {children}
        </CategoriesStoreContext.Provider>
    );
};

const renderWithProviders = (ui: React.ReactElement) => {
    return render(
        <ProductsStoreProvider>
            <TestCategoriesProvider>{ui}</TestCategoriesProvider>
        </ProductsStoreProvider>
    );
};

describe("<MainHome />", () => {
    beforeEach(() => {
        renderWithProviders(<MainHome />);
    });

    it("should render the main landmark with correct accessibility attributes", () => {
        const main = screen.getByRole("main", { name: /catálogo de produtos/i });
        expect(main).toBeInTheDocument();
        expect(main.tagName.toLowerCase()).toBe("main");
    });

    it("should render the Filters section", () => {
        expect(screen.getByTestId("filters-mock")).toBeInTheDocument();
        expect(screen.getByText("Filtros")).toBeVisible();
    });

    it("should render the ProductList", () => {
        const list = screen.getByTestId("product-list-mock");
        expect(list).toBeInTheDocument();
    });


    it("should match snapshot", () => {
        const { container } = renderWithProviders(<MainHome />);
        expect(container).toMatchSnapshot();
    });
});
