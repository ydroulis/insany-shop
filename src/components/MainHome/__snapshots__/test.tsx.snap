// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<MainHome /> should match snapshot 1`] = `
<div>
  <main
    aria-label="Catálogo de produtos"
    class="sc-bRKDuR hLcoMF"
    role="main"
  >
    <div
      data-testid="filters-mock"
    >
      Filtros
    </div>
    <ul
      data-testid="product-list-mock"
    >
      <li>
        <img
          alt="Imagem do produto iPhone 15 Pro"
          src="https://images.unsplash.com/photo-1592750475338-74b7b21085ab?w=400&h=400&fit=crop"
        />
        <span>
          iPhone 15 Pro
        </span>
        <span>
          Eletrônicos
        </span>
        <span>
          R$ 
          8999.99
        </span>
        <span>
          25
           em estoque
        </span>
        <span
          aria-label="Avaliação 4.8"
        >
          ⭐ 
          4.8
        </span>
        <span>
          Apple
        </span>
      </li>
      <li>
        <img
          alt="Imagem do produto Samsung Galaxy S24 Ultra"
          src="https://images.unsplash.com/photo-1610945265064-0e34e5519bbf?w=400&h=400&fit=crop"
        />
        <span>
          Samsung Galaxy S24 Ultra
        </span>
        <span>
          Eletrônicos
        </span>
        <span>
          R$ 
          7499.99
        </span>
        <span>
          18
           em estoque
        </span>
        <span
          aria-label="Avaliação 4.7"
        >
          ⭐ 
          4.7
        </span>
        <span>
          Samsung
        </span>
      </li>
      <li>
        <img
          alt="Imagem do produto MacBook Air M3"
          src="https://images.unsplash.com/photo-1541807084-5c52b6b3adef?w=400&h=400&fit=crop"
        />
        <span>
          MacBook Air M3
        </span>
        <span>
          Eletrônicos
        </span>
        <span>
          R$ 
          12999.99
        </span>
        <span>
          12
           em estoque
        </span>
        <span
          aria-label="Avaliação 4.9"
        >
          ⭐ 
          4.9
        </span>
        <span>
          Apple
        </span>
      </li>
      <li>
        <img
          alt="Imagem do produto PlayStation 5"
          src="https://images.unsplash.com/photo-1606813907291-d86efa9b94db?w=400&h=400&fit=crop"
        />
        <span>
          PlayStation 5
        </span>
        <span>
          Eletrônicos
        </span>
        <span>
          R$ 
          4199.99
        </span>
        <span>
          8
           em estoque
        </span>
        <span
          aria-label="Avaliação 4.6"
        >
          ⭐ 
          4.6
        </span>
        <span>
          Sony
        </span>
      </li>
      <li>
        <img
          alt="Imagem do produto Camiseta Básica Premium"
          src="https://images.unsplash.com/photo-1521572163474-6864f9cf17ab?w=400&h=400&fit=crop"
        />
        <span>
          Camiseta Básica Premium
        </span>
        <span>
          Roupas e Calçados
        </span>
        <span>
          R$ 
          89.99
        </span>
        <span>
          50
           em estoque
        </span>
        <span
          aria-label="Avaliação 4.4"
        >
          ⭐ 
          4.4
        </span>
        <span>
          BasicWear
        </span>
      </li>
      <li>
        <img
          alt="Imagem do produto Jeans Skinny Masculino"
          src="https://images.unsplash.com/photo-1542272604-787c3835535d?w=400&h=400&fit=crop"
        />
        <span>
          Jeans Skinny Masculino
        </span>
        <span>
          Roupas e Calçados
        </span>
        <span>
          R$ 
          199.99
        </span>
        <span>
          30
           em estoque
        </span>
        <span
          aria-label="Avaliação 4.3"
        >
          ⭐ 
          4.3
        </span>
        <span>
          DenimCo
        </span>
      </li>
      <li>
        <img
          alt="Imagem do produto Vestido Floral Feminino"
          src="https://images.unsplash.com/photo-1595777457583-95e059d581b8?w=400&h=400&fit=crop"
        />
        <span>
          Vestido Floral Feminino
        </span>
        <span>
          Roupas e Calçados
        </span>
        <span>
          R$ 
          159.99
        </span>
        <span>
          22
           em estoque
        </span>
        <span
          aria-label="Avaliação 4.5"
        >
          ⭐ 
          4.5
        </span>
        <span>
          FloralStyle
        </span>
      </li>
      <li>
        <img
          alt="Imagem do produto Tênis Esportivo Running"
          src="https://images.unsplash.com/photo-1549298916-b41d501d3772?w=400&h=400&fit=crop"
        />
        <span>
          Tênis Esportivo Running
        </span>
        <span>
          Roupas e Calçados
        </span>
        <span>
          R$ 
          299.99
        </span>
        <span>
          35
           em estoque
        </span>
        <span
          aria-label="Avaliação 4.6"
        >
          ⭐ 
          4.6
        </span>
        <span>
          RunTech
        </span>
      </li>
      <li>
        <img
          alt="Imagem do produto Sofá 3 Lugares Retrátil"
          src="https://images.unsplash.com/photo-1586023492125-27b2c045efd7?w=400&h=400&fit=crop"
        />
        <span>
          Sofá 3 Lugares Retrátil
        </span>
        <span>
          Casa e Decoração
        </span>
        <span>
          R$ 
          1899.99
        </span>
        <span>
          6
           em estoque
        </span>
        <span
          aria-label="Avaliação 4.7"
        >
          ⭐ 
          4.7
        </span>
        <span>
          ComfortHome
        </span>
      </li>
      <li>
        <img
          alt="Imagem do produto Mesa de Jantar 6 Lugares"
          src="https://images.unsplash.com/photo-1549497538-303791108f95?w=400&h=400&fit=crop"
        />
        <span>
          Mesa de Jantar 6 Lugares
        </span>
        <span>
          Casa e Decoração
        </span>
        <span>
          R$ 
          899.99
        </span>
        <span>
          10
           em estoque
        </span>
        <span
          aria-label="Avaliação 4.4"
        >
          ⭐ 
          4.4
        </span>
        <span>
          ModernLiving
        </span>
      </li>
    </ul>
    <section
      aria-labelledby="categories-title"
      class="sc-hvigdm cBLToQ"
      role="region"
    >
      <h2
        class="sc-fhHczv iCxmzm"
        id="categories-title"
      >
        Principais categorias
      </h2>
      <ul
        class="sc-ggWZvA bKXArt"
        role="list"
      >
        <li
          role="listitem"
        >
          <a
            aria-label="Ver categoria Eletrônicos com 4 produtos"
            class="sc-dTvVRJ djcVQG"
            href="/category/eletronicos"
          >
            <article
              class="sc-jwTyAe hTILNK"
              data-testid="category-card"
            >
              <h3
                class="sc-hjsuWn deQKlM"
              >
                Eletrônicos
              </h3>
              <p
                aria-hidden="true"
                class="sc-jJLAfE dEWSat"
              >
                4 produtos
              </p>
            </article>
          </a>
        </li>
        <li
          role="listitem"
        >
          <a
            aria-label="Ver categoria Roupas e Calçados com 4 produtos"
            class="sc-dTvVRJ djcVQG"
            href="/category/roupas"
          >
            <article
              class="sc-jwTyAe hTILNK"
              data-testid="category-card"
            >
              <h3
                class="sc-hjsuWn deQKlM"
              >
                Roupas e Calçados
              </h3>
              <p
                aria-hidden="true"
                class="sc-jJLAfE dEWSat"
              >
                4 produtos
              </p>
            </article>
          </a>
        </li>
        <li
          role="listitem"
        >
          <a
            aria-label="Ver categoria Casa e Decoração com 3 produtos"
            class="sc-dTvVRJ djcVQG"
            href="/category/casa"
          >
            <article
              class="sc-jwTyAe hTILNK"
              data-testid="category-card"
            >
              <h3
                class="sc-hjsuWn deQKlM"
              >
                Casa e Decoração
              </h3>
              <p
                aria-hidden="true"
                class="sc-jJLAfE dEWSat"
              >
                3 produtos
              </p>
            </article>
          </a>
        </li>
        <li
          role="listitem"
        >
          <a
            aria-label="Ver categoria Livros com 3 produtos"
            class="sc-dTvVRJ djcVQG"
            href="/category/livros"
          >
            <article
              class="sc-jwTyAe hTILNK"
              data-testid="category-card"
            >
              <h3
                class="sc-hjsuWn deQKlM"
              >
                Livros
              </h3>
              <p
                aria-hidden="true"
                class="sc-jJLAfE dEWSat"
              >
                3 produtos
              </p>
            </article>
          </a>
        </li>
        <li
          role="listitem"
        >
          <a
            aria-label="Ver categoria Esportes e Lazer com 3 produtos"
            class="sc-dTvVRJ djcVQG"
            href="/category/esportes"
          >
            <article
              class="sc-jwTyAe hTILNK"
              data-testid="category-card"
            >
              <h3
                class="sc-hjsuWn deQKlM"
              >
                Esportes e Lazer
              </h3>
              <p
                aria-hidden="true"
                class="sc-jJLAfE dEWSat"
              >
                3 produtos
              </p>
            </article>
          </a>
        </li>
      </ul>
    </section>
  </main>
</div>
`;
